Data structures via the cli supporting docs (#1017)
snowplow/documentation: docs/recipes/recipe-data-structures-in-git/index.md

On top of that, detailed change history can be crucial for such large organizations. It is important to be able to tell in fine detail what was changed, when, and by whom.

The Snowplow Console's UI offers excellent facilities to get started quickly with Data Structures (either using the Builder or directly the JSON editor), and is a solid tool for smaller teams; but it doesn't implement such approval processes, neither does it offer such fine-grained visibility around changes.

A common solution when faced with these requirements is to move management to some form of version control platform (github/gitlab). This opens up an entire ecosystem of tools and patterns enabling all manner of custom workflows.

We have built [snowplow-cli](/docs/understanding-tracking-design/managing-your-data-structures/cli/) to help you bridge the gap between these repository based workflows and BDP Console.

## Prerequisites

* A deployed Snowplow BDP pipeline
* [snowplow-cli](/docs/understanding-tracking-design/managing-your-data-structures/cli/#download) downloaded and configured

> * A familiarity with [git](https://git-scm.com/) and an understanding of [github workflows](https://docs.github.com/en/actions/writing-workflows)

Reviewer: I think the product is called `github actions`, but Im not sure what's the difference

Author: I thought that too but their docs are a bit confusing. They refer to workflows in a lot of places then actions in their more detailed stuff.

Reviewer: Don't know, need some marketing people. I would use actions since it's a more recognisable name

Author: As far as I can tell, actions is the platform and workflows are the things you create. Yeah, I can change it to actions to avoid anyone who hasn't read the definitions getting confused.

* A sensible [terminal emulator](https://en.wikipedia.org/wiki/Terminal_emulator) and shell

## What you'll be doing

This recipe will walk through creating and deploying a data structure from the command line using [snowplow-cli](https://github.com/snowplow-product/snowplow-cli). It will then show how it is possible to automate the validation and deployment process using [github actions](https://docs.github.com/en/actions/writing-workflows).

## Create a local data structure

Firstly we'll need a place to put things.

```bash
$ mkdir -p snowplow-structures/data-structures
$ cd snowplow-structures
```
:::tip
snowplow-cli data structures commands default to looking for data structures in `./data-structures`.
:::

Now let's create our data structure. We'll create a custom event called 'login'.

```bash
$ snowplow-cli ds generate login --vendor com.example
```
:::note
`ds` is an alias for `data-structures`.
:::

This should provide us the following output

```
3:00PM INFO generate wrote=data-structures/com.example/login.yaml
```

> You'll note that snowplow-cli uses a naming scheme that follows a `vendor/name` pattern and defaults to `yaml` output. This is not a prerequisite for other commands to work, just an opinionated default.

Reviewer 2: This paragraph is unclear to me.

Author: I was trying to get across the pattern but also that you don't have to follow it. Maybe that isn't important to note. I'll reword

data:
  $schema: http://iglucentral.com/schemas/com.snowplowanalytics.self-desc/schema/jsonschema/1-0-0#
  self:
    vendor: com.example
    name: login
    format: jsonschema
    version: 1-0-0
  type: object
  properties:
    result:
      enum: [success, failure]
  additionalProperties: false
```

### Validate

We should validate our changes before we attempt to publish them. Let's do that

```bash
$ snowplow-cli ds validate data-structures/com.example/login.yaml
```
:::tip
You can supply snowplow-cli with a directory and it will look for anything that looks like a data structure. Also given the default data structure directory is being used the previous command is equivalent to `snowplow-cli ds validate`.
:::

You should see output similar to this:
```
3:00PM INFO validating from paths=[data-structures/com.example/login.yaml]
3:00PM INFO will create file=data-structures/com.example/login.yaml vendor=com.example name=login version=1-0-0
3:00PM WARN validation file=data-structures/com.example/login.yaml
  messages=
  │ The schema is missing the "description" property (/properties/result)
  │ The schema is missing the "description" property (/)
```

### Publish to development

> Apart from the missing descriptions everything looks good. We can fill them in later™. Let's go ahead and publish our data structure to our [development](/docs/managing-data-quality/testing-and-qa-workflows/) environment.

Reviewer 2: I found that TM superscript funny but can we please remove it from the docs :)

Author: wrong tone, understood.

![](./images/locked.png)

Any data structures published using snowplow-cli will automatically get this banner and have UI based editing disabled. It is a good idea to settle on one source of truth for each data structure to avoid potential conflicts.

### Publish to production

With our data structure deployed to develop and working as we expect we can safely publish it to production.

```bash
$ snowplow-cli ds publish prod
```
```
3:00PM INFO publishing to prod from paths=[data-structures]
3:00PM INFO will update file=data-structures/com.example/login.yaml local=1-0-0 remote=""
3:00PM INFO all done!
```
:::note
Data structures must be published to `dev` before they can be published to `prod`
:::

We have now seen how to create, validate and then publish a new data structure from the command line. Next we'll look at how to configure github actions to run validation and publishing automatically for us.

## Automating with github actions

### Set up repository

> We'll not go into the details of creating github repositories and initial commits here. The [github docs](https://docs.github.com/) do an excellent job of that already. The next few steps will assume a working github repository containing the directory and data structure we created in the previous section. It will have two branches named `main` and `develop` which should be in sync.

Reviewer 2: ```suggestion
We'll not go into the details of creating github repositories and initial commits here, the [github docs](https://docs.github.com/) provide full guidance how to do those. The next few steps will assume a working github repository containing the directory and data structure we created in the previous section. It will have two branches named `main` and `develop` which should be in sync.
```

Author: At primary school I got roasted for writing extremely long sentences and using commas when I should have used full stops. That roasting has lived with me ever since and now every time I feel like I want to comma I find a way to full stop.

```yml
on:
  push:
    branches: [develop]
```

With our trigger point worked out we need to complete a series of steps:
1. configure snowplow-cli via environment variables provided as [github action secrets](https://docs.github.com/en/actions/security-for-github-actions/security-guides/using-secrets-in-github-actions)
2. checkout our repo
3. install snowplow-cli
4. run the `snowplow-cli ds publish dev` command as we did earlier

The full action:

```yml title=".github/workflows/publish-develop.yml"
on:
  push:
    branches: [develop]

jobs:
  publish:
    runs-on: ubuntu-latest
    env:
      SNOWPLOW_CONSOLE_ORG_ID: ${{ secrets.SNOWPLOW_CONSOLE_ORG_ID }}
      SNOWPLOW_CONSOLE_API_KEY_ID: ${{ secrets.SNOWPLOW_CONSOLE_API_KEY_ID }}
      SNOWPLOW_CONSOLE_API_KEY_SECRET: ${{ secrets.SNOWPLOW_CONSOLE_API_KEY_SECRET }}

    steps:
      - uses: actions/checkout@v4

      - run: curl -L -o snowplow-cli https://github.com/snowplow-product/snowplow-cli/releases/latest/download/snowplow-cli_linux_x86_64 && chmod u+x snowplow-cli

>       - run: ./snowplow-cli ds publish dev --managed-from $GITHUB_REPOSITORY

Reviewer: Should we move the `--managed-from` to be an env variable as well?

Author: maybe? I don't think I feel strongly. It kind of does fit with things like org id, yeah, id be ok with it either way

Reviewer: I think it makes sense, it makes the command a bit cleaner

```
:::tip
The value of the `--managed-from` flag will be displayed inside the 'This data structure is locked' banner we saw earlier in the UI. It is designed to help people track down the source of truth for this data structure.
:::

### Publish to production workflow

In the same way we want our `develop` branch to deploy to our `develop` environment we want our `main` branch to deploy to our `production` environment.

As we saw earlier publishing to production is very similar to publishing to development. The only new thing we need here is a different workflow trigger.

```yml title=".github/workflows/publish-production.yml"
on:
  push:
    branches: [main]

jobs:
  publish:
    runs-on: ubuntu-latest
    env:
      SNOWPLOW_CONSOLE_ORG_ID: ${{ secrets.SNOWPLOW_CONSOLE_ORG_ID }}
      SNOWPLOW_CONSOLE_API_KEY_ID: ${{ secrets.SNOWPLOW_CONSOLE_API_KEY_ID }}
      SNOWPLOW_CONSOLE_API_KEY: ${{ secrets.SNOWPLOW_CONSOLE_API_KEY }}

    steps:
      - uses: actions/checkout@v4

      - run: curl -L -o snowplow-cli https://github.com/snowplow-product/snowplow-cli/releases/latest/download/snowplow-cli_linux_x86_64 && chmod u+x snowplow-cli

>       - run: ./snowplow-cli ds pub prod --managed-from $GITHUB_REPOSITORY

Reviewer 2: `pub`? Not `published`? I assume the arg parse tool may allow that, but let's settle on `publish` as the clearer option.

```

### Validate on pull request workflow

A core component of version control based workflows is the [pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/about-pull-requests). For our repository we would like to ensure as best we can that any data structure changes are valid and problem free before they get merged into develop. Lucky for us there is a github workflow event for that.

By combining the `snowplow-cli ds validate` command and the github workflow pull request event we arrive at this:
```yml title=".github/workflows/validate-pull-request.yml"
on:
  pull_request:
    branches: [develop, main]

jobs:

>   publish:

Reviewer 2: Probably `validate` instead of `publish`?

Having created a [new branch](https://git-scm.com/book/en/v2/Git-Branching-Basic-Branching-and-Merging) called `login-results-error-codes` and making the changes locally we should end up here:
```bash title="git diff develop"
--- a/data-structures/com.example/login.yaml
+++ b/data-structures/com.example/login.yaml
@@ -10,9 +10,9 @@ data:
     vendor: com.example
     name: login
     format: jsonschema
-    version: 1-0-0
+    version: 1-0-1
   type: object
   properties:
     result:
-      enum: [success, failure]
+      enum: [200, 403]
   additionalProperties: false
```

That all looks good so we'll go ahead and push to github and [create a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a-pull-request).

We wait patiently for our validate on pull request workflow to run and then..

![](./images/worked-pr-checks.png)

> Not great. To dig in and find the problem we open the 'file' tab on the pull request and see..

Reviewer 2: ```suggestion
Validation has failed. To identify the problem we open the 'file' tab on the pull request and see..
```

![](./images/worked-diff-annotated.png)
:::note
Validation only takes your configured [destinations](https://console.snowplowanalytics.com/destinations) into account.
:::

Together with the description warnings we forgot to fix earlier we have some errors. Changing the values of the enum would change the type of the `result` property which will cause problems further down the line for our data. The error suggests we need to make a major version bump to avert disaster. We'll do that (and add descriptions).

Our next attempt:
```bash title="git diff develop"
--- a/data-structures/com.example/login.yaml
+++ b/data-structures/com.example/login.yaml
@@ -10,9 +10,11 @@ data:
     vendor: com.example
     name: login
     format: jsonschema
-    version: 1-0-0
+    version: 2-0-0
   type: object
+  description: Login outcome event
   properties:
     result:
-      enum: [success, failure]
+      description: The resulting http error code of a login request
+      enum: [200, 403]
   additionalProperties: false
```

And the workflow result..

![](./images/worked-pr-checks-ok.png)

> Excellent. Now our colleagues can feedback on our changes and if everyone is happy we can merge to `develop` which will trigger our `publish-develop.yml` workflow.

Reviewer 2: ```suggestion
Now that validation has passed, our colleagues can feedback on our changes and if everyone is happy we can merge to `develop` which will trigger our `publish-develop.yml` workflow.
```

![](./images/worked-pub-dev.png)

Finally, once we are convinced everything works we can open another pull request from `develop` to `main`, merge that and trigger our `publish-production.yml` workflow.

## Let's break down what we've done

* We have seen how snowplow-cli can be used to work with data structures from the command line
* We have applied that knowledge to build github workflows which support automated validation and publication

> ## What you might want to do next

Reviewer 2: I would completely remove this section.

Author: Both or just the 'do next'? I was trying to stay consistent with the other recipes but it is a bit redundant.
